relayer/workers/beefyrelayer/store/store.go

```diff
@@ -96,71 +100,78 @@ func NewDatabaseCmd(info *BeefyRelayInfo, cmdType CmdType, instructions map[stri
 }
 
 type Database struct {
+	Path     string
 	DB       *gorm.DB
 	messages <-chan DatabaseCmd
 	log      *logrus.Entry
 }
 
-func NewDatabase(db *gorm.DB, messages <-chan DatabaseCmd, log *logrus.Entry) *Database {
+func NewDatabase(messages <-chan DatabaseCmd, log *logrus.Entry) *Database {
 	return &Database{
-		DB:       db,
+		Path:     "",
+		DB:       nil,
 		messages: messages,
 		log:      log,
 	}
 }
 
-func PrepareDatabase(config *Config) (*gorm.DB, error) {
-	if len(config.DBPath) == 0 {
-		return nil, fmt.Errorf("invalid database path: %s", config.DBPath)
-	}
-	tmpDBFile, err := ioutil.TempFile("", config.DBPath)
+func (d *Database) Initialize() error {
+	tmpfile, err := ioutil.TempFile("", "beefy.*.db")
 	if err != nil {
-		return nil, err
+		return nil
 	}
+	tmpfile.Close()
 
-	db, err := gorm.Open(config.Dialect, tmpDBFile.Name())
+	db, err := gorm.Open("sqlite3", tmpfile.Name())
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	InitTables(db)
-
-	return db, nil
-}
-
-func InitTables(db *gorm.DB) {
 	var beefyRelayInfo BeefyRelayInfo
 	if !db.HasTable(&beefyRelayInfo) {
 		db.CreateTable(&beefyRelayInfo)
 		db.Model(&beefyRelayInfo)
 	}
-}
 
-func (d *Database) onDone(ctx context.Context) error {
-	d.log.Info("Shutting down database...")
-	return ctx.Err()
+	d.Path = tmpfile.Name()
+	d.DB = db
+
+	return nil
 }
 
 func (d *Database) Start(ctx context.Context, eg *errgroup.Group) error {
 	eg.Go(func() error {
-		return d.writeLoop(ctx)
+		var err1, err2 error
+
+		err1 = d.writeLoop(ctx)
+
+		d.log.Info("Shutting down DB")
+		sqlDB := d.DB.DB()
+		if sqlDB != nil {
+			err2 = sqlDB.Close()
+			if err2 != nil {
+				d.log.WithError(err2).Error("Unable to close DB connection")
+			}
+
+			err2 = os.Remove(d.Path)
```

Comment:
nice 👍 - if the relayer crashes and restarts without this, i presume it just erases/overwrites from scratch into a new file when it restarts?